google/datalab/utils/commands/_commands.py

```diff
@@ -63,17 +76,88 @@ def create_args(line, namespace):
         args.append(arg)
     return args
 
-  def parse(self, line, namespace=None):
-    """Parses a line into a dictionary of arguments, expanding meta-variables from a namespace. """
+  def _get_args_to_parse_from_cell(self):
+
+    def _get_subparsers(p):
+      """Recursively get subparsers."""
+
+      subparsers = [
+        subparser
+        for action in p._actions
+        if isinstance(action, argparse._SubParsersAction)
+        for _, subparser in action.choices.items()
+      ]
+      ret = subparsers
+      for sp in subparsers:
+        ret += _get_subparsers(sp)
+      return ret
+
+    if not self._subparser_args_to_parse_from_cell:
+      subparsers = _get_subparsers(self)
+      subcommand_name_to_args = {}
+      for subparser in subparsers:
+        name = subparser.prog.split()[-1]
+        args_to_parse = subparser._args_to_parse_from_cell
+        subcommand_name_to_args[name] = args_to_parse
+      self._subparser_args_to_parse_from_cell = subcommand_name_to_args
+    return self._subparser_args_to_parse_from_cell
+
+  def parse(self, line, cell, namespace=None):
+    """Parses a line and cell into a dictionary of arguments, expanding variables from a namespace.
+
+    For each line parameters beginning with --, it also checks the cell content and see if it exists
+    there. For example, if "--config1" is a line parameter, it checks to see if cell dict contains
+    "config1" item, and if so, use the cell value. The "config1" item will also be removed from
```

Comment:
I think this might cause some misunderstanding. How about we throw an error instead, saying that the parameter is specified in two different ways? Otherwise I'm not sure which should take the precedence.